rslib/src/tags.rs

```diff
@@ -2,16 +2,80 @@
 // License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 
 use crate::{
+    backend_proto::{Tag as TagProto, TagTreeNode},
     collection::Collection,
     err::{AnkiError, Result},
     notes::{NoteID, TransformNoteOutput},
-    text::to_re,
-    {text::normalize_to_nfc, types::Usn},
+    text::{normalize_to_nfc, to_re},
+    types::Usn,
 };
+
 use regex::{NoExpand, Regex, Replacer};
-use std::{borrow::Cow, collections::HashSet};
+use std::cmp::Ordering;
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    iter::Peekable,
+};
 use unicase::UniCase;
 
+#[derive(Debug, Clone)]
+pub struct Tag {
+    pub name: String,
+    pub usn: Usn,
+    pub collapsed: bool,
+}
+
+impl Ord for Tag {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.name.cmp(&other.name)
+    }
```

Comment:
Because the DB stores items in a unicase collation, the tags will arrive in unicode sorted order. But here you're using String::cmp to order, so sorting a vec of tags will change the unicode sort order to simple ascii order. It looks like the missing tag function does a simple sort, so the unicase order would be lost. To prevent that, you'd need to create temporary UniCase objects like in canonify_tags().